benchmarks/python/rope_ops.py

```diff
@@ -772,6 +772,124 @@ def iobytes():
+            # training: flash attention requires it
+            # inference: multi-query would require a full kv cache so avoid it to limit its memory usage
+            if (
+                self.config.n_query_groups != self.config.n_head
+                and self.config.n_query_groups != 1
+            ):
+                k = k.expand(
+                    B, self.config.n_query_groups, q_per_kv, T, self.config.head_size
+                )
+                v = v.expand(
+                    B, self.config.n_query_groups, q_per_kv, T, self.config.head_size
+                )
+
+            q = q.reshape(B, -1, T, self.config.head_size)  # (B, nh_q, T, hs)
+            k = k.reshape(B, -1, T, self.config.head_size)  # (B, nh_k, T, hs)
+            v = v.reshape(B, -1, T, self.config.head_size)  # (B, nh_v, T, hs)
+
+            q_roped = self.apply_rope(q[..., : self.config.rope_n_elem], cos, sin)
+            k_roped = self.apply_rope(k[..., : self.config.rope_n_elem], cos, sin)
+            q = torch.cat((q_roped, q[..., self.config.rope_n_elem :]), dim=-1)
+            k = torch.cat((k_roped, k[..., self.config.rope_n_elem :]), dim=-1)
+            return q, k, v
+
+    cfg = configs["litgpt"]()
+    # overwrite seq_length
+    cfg.seq_len = seq_length
+
+    def inputs():
+        qkv = torch.randn(
+            cfg.batch_size,
+            cfg.seq_len,
+            (cfg.n_head + 2 * cfg.n_query_groups) * cfg.head_size,
+            device="cuda",
+            dtype=torch.bfloat16,
+            requires_grad=True,
+        )
+        cos = torch.randn(
+            1,
+            cfg.seq_len,
+            cfg.rope_n_elem,
+            device="cuda",
+            dtype=torch.bfloat16,
+            requires_grad=False,
+        )
+        sin = torch.randn(
+            1,
+            cfg.seq_len,
+            cfg.rope_n_elem,
+            device="cuda",
+            dtype=torch.bfloat16,
+            requires_grad=False,
+        )
+        return qkv, cos, sin
+
+    def grads():
+        grad = torch.randn(
+            cfg.batch_size,
+            cfg.n_head,
+            cfg.seq_len,
+            cfg.head_size,
+            device="cuda",
+            dtype=torch.bfloat16,
+            requires_grad=False,
+        )
+        return grad
+
+    # Manual IOBytes computes the total bandwidth for thunder backward trace.
+    def iobytes():
+        n_elements = 1
+        # adding size of qkv.grad
+        n_elements += (
+            cfg.batch_size * cfg.seq_len * (cfg.n_head + 2 * cfg.n_query_groups) * cfg.head_size,
+        )
+        # adding size of sin, cos (saved from forward)
+        n_elements += (
+            2 * cfg.seq_len * cfg.rope_n_elem
+        )
+        # adding size of q, k, v (saved from forward)
+        n_elements += (
+            3 * cfg.batch_size * cfg.num_attention_heads * cfg.seq_len * cfg.head_dim
+        )
+        # totoal io sizes
+        return n_elements * torch.bfloat16.itemsize
```

Comment:
Should this be using dtype instead of `torch.bfloat16`?

Reply:
I guess we've been running bf16 only in rope benchmark the whole time. (so we don't have a `dtype` arg in these tests.

I don't think we necessarily need to expand this to fp32?!

Reply:
I don't have a strong preference though, so I'll let you choose whatever is most convenient for you. Using the `dtype` directly has the potential of a mismatch in different parts of the model in case of future updates.

Reply:
I'll follow up with a small PR converging the dtype across these `rope_ops.py`